### ISSUE TITLE
cpan App::perlbrew found a fake-bin curl instead of curl?

### ISSUE THREAD

**Issue by dlamblin** [gugod/App-perlbrew#599]
I may not be reading the output right, but did it really just fail because the `curl` it chose was in `App-perlbrew-0.82-…/fake-bin/curl` ?

```
LamblinCL5116:~ lamblin$ cpan App::perlbrew

CPAN.pm requires configuration, but most of it can be done automatically.
If you answer 'no' below, you will enter an interactive dialog for each
configuration option instead.

Would you like to configure as much as possible automatically? [yes]
Use of uninitialized value $what in concatenation (.) or string at /System/Library/Perl/5.18/App/Cpan.pm line 553, <STDIN> line 1.

Warning: You do not have write permission for Perl library directories.

To install modules, you need to configure a local Perl library directory or
escalate your privileges.  CPAN can help you by bootstrapping the local::lib
module or by configuring itself to use 'sudo' (if available).  You may also
resolve this problem manually if you need to customize your setup.

What approach do you want?  (Choose 'local::lib', 'sudo' or 'manual')
 [local::lib]

Would you like me to automatically choose some CPAN mirror
sites for you? (This means connecting to the Internet) [yes]
Attempting to create directory /Users/lamblin/perl5

Checking if your kit is complete...
Looks good
Generating a Unix-style Makefile
Writing Makefile for local::lib
Writing MYMETA.yml and MYMETA.json
cp lib/lib/core/only.pm blib/lib/lib/core/only.pm
cp lib/local/lib.pm blib/lib/local/lib.pm
cp lib/POD2/PT_BR/local/lib.pod blib/lib/POD2/PT_BR/local/lib.pod
cp lib/POD2/DE/local/lib.pod blib/lib/POD2/DE/local/lib.pod
Manifying 4 pod documents
PERL_DL_NONLAZY=1 "/usr/bin/perl" "-I/Users/lamblin/perl5/lib/perl5" "-MExtUtils::Command::MM" "-MTest::Harness" "-e" "undef *Test::Harness::Switches; test_harness(0, 'blib/lib', 'blib/arch')" t/*.t
t/bad_variables.t ...... ok
t/carp-mismatch.t ...... ok
t/classmethod.t ........ ok
t/de-dup.t ............. ok
t/lib-core-only.t ...... ok
t/pipeline.t ........... ok
t/shell.t .............. ok
t/stackable.t .......... ok
t/subroutine-in-inc.t .. ok
t/taint-mode.t ......... ok
All tests successful.
Files=10, Tests=220,  5 wallclock secs ( 0.08 usr  0.02 sys +  2.29 cusr  1.45 csys =  3.84 CPU)
Result: PASS
Manifying 4 pod documents
Installing /Users/lamblin/perl5/lib/perl5/POD2/PT_BR/local/lib.pod
Installing /Users/lamblin/perl5/lib/perl5/POD2/DE/local/lib.pod
Installing /Users/lamblin/perl5/lib/perl5/local/lib.pm
Installing /Users/lamblin/perl5/lib/perl5/lib/core/only.pm
Installing /Users/lamblin/perl5/man/man3/lib::core::only.3pm
Installing /Users/lamblin/perl5/man/man3/POD2::DE::local::lib.3pm
Installing /Users/lamblin/perl5/man/man3/local::lib.3pm
Installing /Users/lamblin/perl5/man/man3/POD2::PT_BR::local::lib.3pm
Appending installation info to /Users/lamblin/perl5/lib/perl5/darwin-thread-multi-2level/perllocal.pod
Running install for module 'App::perlbrew'
Running make for G/GU/GUGOD/App-perlbrew-0.82.tar.gz
Fetching with LWP:
http://mirror.bibleonline.ru/cpan/authors/id/G/GU/GUGOD/App-perlbrew-0.82.tar.gz
Fetching with LWP:
http://mirror.bibleonline.ru/cpan/authors/id/G/GU/GUGOD/CHECKSUMS
Checksum for /Users/lamblin/.cpan/sources/authors/id/G/GU/GUGOD/App-perlbrew-0.82.tar.gz ok

  CPAN.pm: Building G/GU/GUGOD/App-perlbrew-0.82.tar.gz

Checking if your kit is complete...
Looks good
Generating a Unix-style Makefile
Writing Makefile for App::perlbrew
Writing MYMETA.yml and MYMETA.json
cp lib/App/perlbrew.pm blib/lib/App/perlbrew.pm
cp bin/perlbrew blib/script/perlbrew
"/usr/bin/perl" "-Iinc" -MExtUtils::MY -e 'MY->fixin(shift)' -- blib/script/perlbrew
Manifying 1 pod document
Manifying 1 pod document
  GUGOD/App-perlbrew-0.82.tar.gz
  /usr/bin/make -- OK
Running make test
PERL_DL_NONLAZY=1 "/usr/bin/perl" "-MExtUtils::Command::MM" "-MTest::Harness" "-e" "undef *Test::Harness::Switches; test_harness(0, 'inc', 'blib/lib', 'blib/arch')" t/*.t
t/00.load.t ............................ ok
t/01.options.t ......................... ok
t/02.format_perl_version.t ............. ok
t/03.path_with_tilde.t ................. ok
t/03.test_get_available_versions.t ..... ok
t/04.find_available_perls.t ............ ok
t/05.get_current_perl.t ................ ok
t/06.installed_perls.t ................. skipped: No perl installation under PERLBREW_ROOT
t/07.argv.t ............................ ok
t/08.error_available.t ................. ok
t/08.error_install.t ................... ok
t/08.error_install_blead.t ............. ok
t/08.error_install_cpanm.t ............. ok
t/09.exit_status.t ..................... ok
t/10.resolve.t ......................... ok
t/11.editdist.t ........................ ok
t/11.root_from_arg.t ................... ok
t/12.destdir.t ......................... ok
t/12.sitecustomize.t ................... ok
t/13.perl_release.t .................... ok
t/14.perl_version_parsing.t ............ ok
t/15.no-reuse-build-dir.t .............. 1/6 # /var/folders/gd/q65byyhs54g3qknbyv5r7q8r0000gp/T/ZPxEDWaEAr/build/test/test
t/15.no-reuse-build-dir.t .............. ok
t/16.release-detail.t .................. ok
t/17.release-detail-cperl-local.t ...... ok
t/17.release-detail-cperl-remote.t ..... skipped: These tests send HTTP requests. Set env TEST_LIVE=1 to really run them.
t/18.release-detail-perl-local.t ....... ok
t/18.release-detail-perl-remote.t ...... skipped: These tests send HTTP requests. Set env TEST_LIVE=1 to really run them.
t/command-alias.t ...................... ok
t/command-available.t .................. ok
t/command-compgen.t .................... ok
t/command-display-rc.t ................. ok
t/command-env.t ........................ 1/? Attempting to create directory /var/folders/gd/q65byyhs54g3qknbyv5r7q8r0000gp/T/BIfhWGJuIr/libs/perl-5.14.1@nobita
t/command-env.t ........................ ok
t/command-exec.t ....................... 1/? Command [perl -E 'somesub 42'] terminated with exit code 7 ($? = 1792) under the following perl environment:
format_info_output_value
Command [perl -E 'say 42'] terminated with exit code 3 ($? = 768) under the following perl environment:
Command [perl -E 'say 42'] terminated with exit code 3 ($? = 768) under the following perl environment:
Command [perl -E 'say 42'] terminated with exit code 3 ($? = 768) under the following perl environment:
Command [perl -E 'say 42'] terminated with exit code 255 ($? = -1) under the following perl environment:
Command [perl -E 'say 42'] terminated with exit code 7 ($? = 1792) under the following perl environment:
t/command-exec.t ....................... ok
t/command-help.t ....................... ok
t/command-info.t ....................... ok
t/command-install-cpanm.t .............. ok
t/command-install-patchperl.t .......... ok
t/command-lib.t ........................ ok
t/command-list.t ....................... ok
t/current_perl.t ....................... ok
t/do_system.t .......................... ok
t/failure-command-install-cpanm.t ...... ok
t/failure-command-install-patchperl.t .. ok
t/http-program-control.t ............... ok
t/http-ua-detect-non-curl.t ............ # PATH=/Users/lamblin/.cpan/build/App-perlbrew-0.82-XaVdXO/t/fake-bin:/Users/lamblin/perl5/bin:/Library/Java/JavaVirtualMachines/jdk1.8.0_144.jdk/Contents/Home/bin:/Users/lamblin/gohdfs-v1.1.0-darwin-amd64/:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin
# curl = /Users/lamblin/.cpan/build/App-perlbrew-0.82-XaVdXO/t/fake-bin/curl
t/http-ua-detect-non-curl.t ............ 1/? [ERROR] Cannot find a proper http user agent program. Please install curl or wget.
# Tests were run but no plan was declared and done_testing() was not seen.
# Looks like your test exited with 2 just after 1.
t/http-ua-detect-non-curl.t ............ Dubious, test returned 2 (wstat 512, 0x200)
All 1 subtests passed
t/http-ua-detect.t ..................... ok
t/http.t ............................... skipped: (no reason given)
t/installation-perlbrew.t .............. ok
t/installation.t ....................... ok
t/installation2.t ...................... ok
t/installation3.t ...................... ok
t/unit-files-are-the-same.t ............ ok

Test Summary Report
-------------------
t/http-ua-detect-non-curl.t          (Wstat: 512 Tests: 1 Failed: 0)
  Non-zero exit status: 2
  Parse errors: No plan found in TAP output
Files=52, Tests=319,  9 wallclock secs ( 0.22 usr  0.12 sys +  6.63 cusr  1.75 csys =  8.72 CPU)
Result: FAIL
Failed 1/52 test programs. 0/319 subtests failed.
make: *** [test_dynamic] Error 255
  GUGOD/App-perlbrew-0.82.tar.gz
  /usr/bin/make test -- NOT OK
//hint// to see the cpan-testers results for installing this module, try:
  reports GUGOD/App-perlbrew-0.82.tar.gz
Running make install
  make test had returned bad status, won't install without force
LamblinCL5116:~ lamblin$ which curl
/usr/bin/curl
```

**Comment by gugod** [comment]
In a sense... yes. This fake curl program was added to trigger the code path that would return "wget" as the UA program. It looks like the the test requires some update now to also cover the case when "wget" is missing or when it fails to response `wget --version`

**Comment by petersmax** [comment]
Any updates on a remedy for this?

**Comment by gugod** [comment]
This was addressed in PR #645
